File: src/ruby/tormenta20/models/raca.rb
# frozen_string_literal: true

require "active_record"

module Tormenta20
  module Models
    # ActiveRecord model for raças (playable races).
    #
    # @example Find all races
    #   Raca.all
    #
    # @example Get the attribute bonus for a specific attribute
    #   Raca.find_by(id: "elfo").attribute_bonus_for("DES")  # => 2
    #
    # @!attribute [rw] id
    #   @return [String] Unique identifier (e.g. +"humano"+, +"elfo"+)
    # @!attribute [rw] name
    #   @return [String] Race name
    # @!attribute [rw] description
    #   @return [String, nil] Lore description
    # @!attribute [rw] size
    #   @return [String] Size category — see {SIZES}
    # @!attribute [rw] movement
    #   @return [Integer] Base movement speed in metres
    # @!attribute [rw] vision
    #   @return [String] Vision type — see {VISIONS}
    # @!attribute [rw] vision_range
    #   @return [Integer, nil] Range of special vision in metres, or +nil+ for normal vision
    # @!attribute [rw] attribute_bonuses
    #   @return [Hash{String => Integer}] Map of attribute key to bonus value
    # @!attribute [rw] skill_bonuses
    #   @return [Array] Skill bonus descriptors
    # @!attribute [rw] racial_abilities
    #   @return [Array<String>] IDs of abilities all members of this race possess
    # @!attribute [rw] chosen_abilities_amount
    #   @return [Integer] Number of abilities the player selects at character creation
    # @!attribute [rw] available_chosen_abilities
    #   @return [Array] Pool of ability IDs to choose from
    class Raca < Base
      self.table_name = "racas"

      SIZES = %w[minúsculo pequeno médio grande].freeze
      VISIONS = %w[normal baixa_luminosidade visao_no_escuro].freeze

      validates :id, presence: true, uniqueness: true
      validates :name, presence: true
      validates :size, inclusion: { in: SIZES }
      validates :movement, numericality: { only_integer: true, greater_than: 0 }

      # Returns the racial attribute bonus for the given attribute key.
      # @param attribute [String, Symbol] Attribute key (e.g. +"DES"+, +"FOR"+)
      # @return [Integer] Bonus value, or +0+ if no bonus exists
      def attribute_bonus_for(attribute)
        attribute_bonuses&.dig(attribute.to_s) || 0
      end

      # @return [Boolean] +true+ if the race is Tiny size
      def minusculo?
        size == "minúsculo"
      end

      # @return [Boolean] +true+ if the race is Small size
      def pequeno?
        size == "pequeno"
      end

      # @return [Boolean] +true+ if the race is Large size
      def grande?
        size == "grande"
      end

      # @return [Boolean] +true+ if the race has darkvision
      def visao_no_escuro?
        vision == "visao_no_escuro"
      end

      def to_h
        {
          id: id,
          name: name,
          description: description,
          size: size,
          movement: movement,
          vision: vision,
          vision_range: vision_range,
          attribute_bonuses: attribute_bonuses || {},
          skill_bonuses: skill_bonuses || [],
          racial_abilities: racial_abilities || [],
          chosen_abilities_amount: chosen_abilities_amount || 0,
          available_chosen_abilities: available_chosen_abilities || []
        }.compact
      end
    end
  end
end

File: src/ts/models/raca.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"
import { getBookReference, type BookReference } from "./concerns/book_referenceable.js"

const TABLE = "racas"

/** Size category of a race. */
export type RacaSize = "minúsculo" | "pequeno" | "médio" | "grande"

/** Vision type of a race. */
export type RacaVision = "normal" | "baixa_luminosidade" | "visao_no_escuro"

/**
 * A Raça (playable race) in Tormenta20.
 *
 * @example
 * ```ts
 * Raca.all()
 * Raca.find("humano")
 *
 * const elfo = Raca.find("elfo")!
 * elfo.attributeBonusFor("DES")   // +2
 * elfo.hasVisaoNoEscuro           // false
 * ```
 */
export class Raca extends BaseModel {
  get description(): string { return this._row.description as string }
  get size(): RacaSize { return this._row.size as RacaSize }
  /** Base movement speed in metres. */
  get movement(): number { return this._row.movement as number }
  get vision(): RacaVision { return this._row.vision as RacaVision }
  /** Effective range of darkvision/low-light vision in metres, or `null` for normal vision. */
  get visionRange(): number | null { return (this._row.vision_range as number) ?? null }
  /** Map of attribute key → bonus value (e.g. `{ DES: 2, INT: -2 }`). */
  get attributeBonuses(): Record<string, number> { return BaseModel.json(this._row.attribute_bonuses, {}) }
  /** JSON array of skill bonus descriptors. */
  get skillBonuses(): unknown[] { return BaseModel.json(this._row.skill_bonuses, []) }
  /** IDs of racial abilities all members of this race have. */
  get racialAbilities(): string[] { return BaseModel.json(this._row.racial_abilities, []) }
  /** Number of racial abilities the player may choose at character creation. */
  get chosenAbilitiesAmount(): number { return (this._row.chosen_abilities_amount as number) ?? 0 }
  /** Pool of ability IDs the player chooses from. */
  get availableChosenAbilities(): unknown[] { return BaseModel.json(this._row.available_chosen_abilities, []) }

  /**
   * Returns the racial attribute bonus for the given attribute key (e.g. `"FOR"`, `"DES"`).
   * Returns `0` if there is no bonus for that attribute.
   */
  attributeBonusFor(attribute: string): number {
    return this.attributeBonuses[attribute] ?? 0
  }

  get isMinusculo(): boolean { return this.size === "minúsculo" }
  get isPequeno(): boolean { return this.size === "pequeno" }
  get isGrande(): boolean { return this.size === "grande" }
  /** `true` if the race has darkvision. */
  get hasVisaoNoEscuro(): boolean { return this.vision === "visao_no_escuro" }

  /** Returns the sourcebook page reference, or `null` if not indexed. */
  bookReference(): BookReference | null { return getBookReference(this.id) }

  toH() {
    return {
      id: this.id, name: this.name, description: this.description,
      size: this.size, movement: this.movement, vision: this.vision,
      vision_range: this.visionRange, attribute_bonuses: this.attributeBonuses,
      skill_bonuses: this.skillBonuses, racial_abilities: this.racialAbilities,
      chosen_abilities_amount: this.chosenAbilitiesAmount,
      available_chosen_abilities: this.availableChosenAbilities,
    }
  }

  static query(): Query<Raca> { return BaseModel.makeQuery(TABLE, Raca) }
  static all(): Raca[] { return Raca.query().all() }
  static find(id: string): Raca | null { return Raca.query().find(id) }
  static first(): Raca | null { return Raca.query().first() }
  static last(): Raca | null { return Raca.query().last() }
  static count(): number { return Raca.query().count() }
}
